File: MergeSort.c
#include <stdio.h>

void merge(int l, int h, int mid , int arr[]){
    int i = l;
    int j = mid+1;
    int k = l;
    int A[100];
    
    while(i<= mid && j<=h){
        if(arr[i]<arr[j]){
            A[k]= arr[i];
            i++;
            k++;
        }

        else{
            A[k] = arr[j];
            j++;
            k++;
        }
    }
    while(i<=mid){
        A[k]= arr[i];
        i++;
        k++;
    }

    while(j<=h){
        A[k]= arr[j];
        j++;
        k++;
    }

    for(int i = 0; i <= h; i++){
        arr[i] = A[i];
    }

}

void mergeSort(int arr[], int l, int h)
{
    if (l < h)
    {
        int mid = (l + h) / 2;
        mergeSort(arr, l, mid);     // sort the left side
        mergeSort(arr, mid + 1, h); // sort the right side
        merge(l, h, mid, arr);
    }
}

void printArray(int *A, int size)
{
    for (int i = 0; i < size; i++)
    {
        printf("%d ", A[i]);
    }
    printf("\n");
}


int main(){
    int arr[] ={2,6,9,4,5,6,4};
    mergeSort(arr,0,6);
    printArray(arr, 7);
        return 0;
 }

File: Difficulty: Medium/Merge Sort/merge-sort.java
class Solution {
    void merge(int arr[], int lt, int r, int mid){
        ArrayList<Integer> list = new ArrayList<>();
        int l = lt;
        int i = mid + 1;
        
        while(l <= mid && i <= r){
            if(arr[l] <= arr[i]){
                list.add(arr[l++]);
            } else {
                list.add(arr[i++]);
            }
        }
        
        while(l <= mid){
            list.add(arr[l++]);
        }
        
        while(i <= r){
            list.add(arr[i++]);
        }
        
        for (int j = 0; j < list.size(); j++) {
            arr[lt + j] = list.get(j);
        }
    }

    void mergeSort(int arr[], int l, int r) {
        if(l >= r){
            return;
        }
        int mid = l + (r - l) / 2;
        mergeSort(arr, l, mid);
        mergeSort(arr, mid + 1, r);
        merge(arr, l, r, mid);
    }
}
